Automate hub deploys when merging to master

scottrigby opened the pull request:
Fixes #43 

- Fixed AZ service principal issue
- CircleCI config
- Move helm-hub-values from keybase repo to public config dir
- Simple azure-cli-helm Dockerfile

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby commented:
@mattfarina I fixed the service principal thing finally, after fussing with it over the weekend. Also set up the job in circleci, except didn't add a value to `SERVICE_PRINCIPAL_PASS` yet, until at least another set of eyes are on this PR. Until then, the job will just fail.

What to do with the `Dockerfile` is another fun question - keeping all in one place for now. If this seems like a good place to stay (I think it could be, unless we think we'll re-use it) then we could hook it up to automatically build on dockerhub. I haven't set that up yet but could if we want once merged. Or we could move the file to another git repo somewhere. For now pushed to my docker hub user.

scottrigby commented:
I'd love to use existing images as opposed to an all-in-one, so sketched that up in a separate branch [here](https://github.com/scottrigby/hub/compare/automate-hub-updates...scottrigby:automate-hub-updates-workflows). Would like to get some feedback though before playing with that more. Maybe we can start with one image and iterate?

mattfarina commented on .circleci/config.yaml:
```diff
@@ -0,0 +1,36 @@
+      - deploy:
+          filters:
+            branches:
+              only: master
```
This is going to run on every change to the repo including docs updates and things unrelated to the config. Can we have a script check for just config changes that should be deployed?

scottrigby replied:
Agree, just for any change in one of the three YAML files 👍

scottrigby replied:
This should now be checking changes to only `config/helm-hub-values.yaml`, `config/repo-values.yaml` and `repos.yaml`.

scottrigby replied:
- [x] Tested on my fork

mattfarina commented on .circleci/config.yaml:
```diff
@@ -0,0 +1,36 @@
+            helm init --client-only
+            helm repo add monocular https://helm.github.io/monocular
+            helm repo update
+            helm upgrade --install helm-hub monocular/monocular --version $CHART_VERSION --namespace helm-hub --set mongodb.mongodbRootPassword=$MONGODB_ROOT_PASSWORD -f config/helm-hub-values.yaml -f config/repo-values.yaml
```
What happens if 2 PRs are merged to master in a short period and the one that is not the tip of master finishes last? The deploy will not contain the latest change merged. How can we prevent that?

scottrigby replied:
That's true. I guess we could go back to the raw git file URLs from master?

prydonius replied:
That's a very good catch, what about instead of doing this for every push to master, we instead run this job on a schedule (e.g. every 15m). We'd need some way to figure out if something changed though, perhaps we can diff the values?

jlegrone replied:
This might do the trick? https://circleci.com/docs/2.0/skip-build/#auto-cancelling-a-redundant-build

scottrigby replied:
@jlegrone given this caveat:
>Note: Your project’s default branch will never auto-cancel builds. This feature only applies to non-workflow builds, builds triggered by pushes to GitHub, or workflow builds that use the new pipelines feature.

I suppose we could enable the pipeline feature, to allow auto-cancelling redundant builds. In fact I've used an orb successfully (that requires pipelines anyway), which would help with [this comment about only diffing the configs](https://github.com/helm/hub/pull/70#discussion_r279406832) (@prydonius that relates to your comment above - we could cron, or we could just diff the configs on pushes to master using the `CIRCLE_COMPARE_URL` from that orb?).

scottrigby replied:
Added the orb. 
- [ ] To-do: will need to enable pipelines in circle once this PR is approved and before this is merged.

scottrigby replied:
Tested on my fork 👍

scottrigby replied:
Pipelines and auto-canceling redundant builds enabled 👍

scottrigby commented on .circleci/Dockerfile:
```diff
@@ -0,0 +1,9 @@
+
+FROM lachlanevenson/k8s-kubectl:v1.14.1
+
+FROM microsoft/azure-cli:latest
```
I was thinking the same. According to their service principal docs for auth, we should use
>the Azure CLI version 2.0.61 or later
but their tagged versions stop before this.

Initially I wrote an alpine 3.9 Dockerfile for the azure-cli, but was trying to stay minimal with what we maintain. Now that you mention it, maybe I should go back to this (there are [some](https://github.com/Azure/azure-cli/issues/8863) [issues](https://github.com/Azure/azure-cli/issues/9167) that solve for this which I referred to). This will likely fix those CVEs.

scottrigby replied:
Image tagged, CVEs [fixed](https://quay.io/repository/helmpack/kubectl-helm-az/manifest/sha256:f880d3797da1c914a7bedf2674741d6b2ad2b07c0c7c8caa16efcb15f3d49a96?tab=vulnerabilities) 👍

scottrigby commented on .circleci/config.yaml:
```diff
@@ -0,0 +1,36 @@
+    docker:
+      # Upgrade the Dockerfile and push image when the hub az cluster tiller
+      # version upgrades.
+      - image: r6by/kubectl-helm-azure
```
Absolutely. Wanted to get feedback on this before going ahead with that. I can make a new repo for the image (something like `helm/kubectl-helm-azure`, and auto push to quay.io/repository/helmpack/kubectl-helm-azure.

scottrigby replied:
Added to Quay, [tests passing](https://quay.io/repository/helmpack/kubectl-helm-az/manifest/sha256:f880d3797da1c914a7bedf2674741d6b2ad2b07c0c7c8caa16efcb15f3d49a96?tab=vulnerabilities). GItHub project here: https://github.com/helm/docker-kubectl-helm-az

prydonius commented on .circleci/Dockerfile:
```diff
@@ -0,0 +1,9 @@
+# Upgrade the helm client version here when the az cluster tiller version
```
this is a pretty clever dockerfile ;)

scottrigby replied:
Heh thx 😄 For future ref, Dockerfile moved to https://github.com/helm/docker-kubectl-helm-az 🔍🕵️‍♂️

scottrigby commented:
I think I've addressed all the concerns so far. Anyone interested to see this test out? I'll wait to add the `SERVICE_PRINCIPAL_PASS` ENV var to the CircleCI project until we're ready for that.

mattfarina commented on .circleci/config.yml:
```diff
@@ -0,0 +1,58 @@
+            if [ -n "$CONFIG_UPDATED" ]; then
+              export MONGODB_ROOT_PASSWORD=$(kubectl get secret --namespace helm-hub helm-hub-mongodb -o jsonpath="{.data.mongodb-root-password}" | base64 -d)
+              helm init --client-only
+              helm repo add monocular https://helm.github.io/monocular
```
In this container the user runs as root. I'm having trouble running `helm init` and `helm repo add` to add the repository. It fails with errors and isn't setting up the right files in `/root/.helm/`.

scottrigby replied:
Ah yes, I didn't add `ca-certificates` - good catch. Rolled new image tag and bumped version.

scottrigby commented:
Following up from https://github.com/helm/hub/pull/70#issuecomment-487421324 I just added `SERVICE_PRINCIPAL_PASS` to circle.
